NDTensors/src/lib/Sectors/src/abstractcategory.jl

```diff
@@ -32,14 +19,100 @@ function Base.show(io::IO, cs::Vector{<:AbstractCategory})
+end
+
+function dimension(g::GradedAxes.GradedUnitRange)
+  return sum(
+    LabelledNumbers.unlabel(b) * dimension(LabelledNumbers.label(b)) for
+    b in blocklengths(g)
+  )
+end
+
+function GradedAxes.dual(category_type::Type{<:AbstractCategory})
   return error("`dual` not defined for type $(category_type).")
 end
 
-Base.isless(c1::AbstractCategory, c2::AbstractCategory) = isless(label(c1), label(c2))
+# ================  fuion rule interface ====================
+function label_fusion_rule(category_type::Type{<:AbstractCategory}, l1, l2)
+  return error("`label_fusion_rule` not defined for type $(category_type).")
+end
+
+# TBD always return GradedUnitRange?
+function fusion_rule(c1::C, c2::C) where {C<:AbstractCategory}
+  out = label_fusion_rule(C, label(c1), label(c2))
+  if typeof(out) <: Tuple{Vector,Vector}  # TODO replace with Trait
+    degen, labels = out
+    # NonAbelianGroup or NonGroupCategory: return GradedUnitRange
+    return GradedAxes.gradedrange(LabelledNumbers.LabelledInteger.(degen, C.(labels)))
+  end
+  return C(out)  # AbelianGroup: return Category
+end
+
+function fusion_rule(g::GradedAxes.GradedUnitRange, c::AbstractCategory)
+  return fusion_rule(c, g)
+end
+
+function ⊗(c1::C, c2::C) where {C<:AbstractCategory}
+  return fusion_rule(c1, c2)
+end
+
+# =============  fusion rule and gradedunitrange ===================
+# TBD define ⊗(c, g2), ⊗(g1, c), ⊗(g1, g2)?
+function GradedAxes.tensor_product(
+  g1::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}}, c::C
+) where {V,C<:AbstractCategory}
+  g2 = gradedrange(c)
+  return GradedAxes.tensor_product(g1, g2)
+end
+
+function GradedAxes.tensor_product(
+  c::C, g::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}}
+) where {V,C<:AbstractCategory}
+  return c ⊗ g
+end
+
+function GradedAxes.tensor_product(
+  g1::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}},
+  g2::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}},
+) where {V,C<:AbstractCategory}
+  return g1 ⊗ g2
+end
+
+GradedAxes.fuse_labels(c1::AbstractCategory, c2::AbstractCategory) = c1 ⊗ c2
+
+# ===============  sum rules ====================
+⊕(c1::C, c2::C) where {C<:AbstractCategory} = GradedAxes.gradedrange([c1 => 1, c2 => 1])
+
+function ⊕(
+  c::C, g::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}}
+) where {V<:Integer,C<:AbstractCategory}
+  return GradedAxes.gradedrange([c => 1]) ⊕ g
+end
+
+function ⊕(
+  g::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}}, c::C
+) where {V<:Integer,C<:AbstractCategory}
+  return g ⊕ GradedAxes.gradedrange([c => 1])
+end
+
+function ⊕(
+  g1::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}},
+  g2::GradedAxes.GradedUnitRange{Vector{LabelledNumbers.LabelledInteger{V,C}}},
+) where {V<:Integer,C<:AbstractCategory}
+  return GradedAxes.chain(g1, g2)
+end
```

Comment:
I think all of these type signatures are too restrictive, and we can just use `AbstractUnitRange` instead of `GradedUnitRange`.

It may not make sense to direct sum certain `AbstractUnitRange` objects with an `AbstractCategory`, but I think we can think about a different way to catch that kind of issue, for example perhaps that is caught in `chain`/`axis_cat`, or we have an orthogonal concept of axis promotion, for example:
```julia
function ⊕(g::AbstractUnitRange c::AbstractCategory)
  return ⊕(gradedaxis_promote(g, c)...)
end

function gradedaxis_promote(g::GradedUnitRange, c::AbstractCategory)
  return (g, gradedrange([c => 1]))
end

function gradedaxis_promote(g::AbstractUnitRange, c::AbstractCategory)
  return error("No promotion defined.")
end
```
That kind of promotion system is how Julia handles algebra between various number types. We probably want a lower level concept of an `gradedaxis_promote_rule` that handles both:
```julia
gradedaxis_promote(g::GradedUnitRange, c::AbstractCategory)
gradedaxis_promote(c::AbstractCategory, g::GradedUnitRange)
```
in a single `gradedaxis_promote_rule`, like how Julia has `promote_type` which calls `promote_rule` internally.

Reply:
This does make me wonder how often we need to direct sum a category onto a graded axis, so perhaps there is another way of thinking about this concept. I guess this is used in generic fusion code, where results of fusions need to be direct summed with each other to make the full space, but fusions can output either a category or a graded unit range?

If that is the case, it seems like one way to handle this would be to not define `⊕(g::GradedUnitRange c::AbstractCategory)` at all (which feels like a strange thing to define, from a generic coding perspective) and instead, in the fusion code, eagerly promote to a `GradedUnitRange` _before_ things get input into `⊕`.

So let's think about the code flow and the context for where these definitions are being used.

Reply:
Here I updated previous code by @emstoudenmire . We do not really need to direct sum spaces, at least not for the symmetric tensor network implementation I have in mind. However I think we can be slightly more ambitious and implement group theory in a generic way.
Anyway promoting looks like the way to go as it also be used for fusion.

Reply:
I see, thanks for the clarification. Agreed, many of the promotion rules should be the same between this code and fusion. I would lean toward removing `⊕` from this package for now. For the time being we can use `axis_cat` as needed, and add back `⊕` later on as a nicer interface on top of `axis_cat`.

Reply:
Finally promotion is mostly avoided by specializing according to `SymmetryStyle`. The few cases left are deal with `to_graded_axis`.